### ISSUE TITLE
Is there an option / attribute make this work as intended?

### ISSUE THREAD

**Issue by AdamSpeight2008** [dotnet/roslyn#18769]
This is likely a `As Designed` but we should have an option to make this work.
[ConsoleApp2.zip](https://github.com/dotnet/roslyn/files/927076/ConsoleApp2.zip)
**Version Used**: 
```vbnet
Microsoft Visual Studio Enterprise 2017 
Version 15.1 (26403.7) Release
VisualStudio.15.Release/15.1.0+26403.7
Microsoft .NET Framework
Version 4.6.01586

Installed Version: Enterprise

Architecture Diagrams and Analysis Tools   00369-60000-00001-AA571
Microsoft Architecture Diagrams and Analysis Tools

Visual Basic 2017   00369-60000-00001-AA571
Microsoft Visual Basic 2017

Visual C# 2017   00369-60000-00001-AA571
Microsoft Visual C# 2017

Application Insights Tools for Visual Studio Package   8.6.00209.10
Application Insights Tools for Visual Studio

ASP.NET and Web Tools 2017   15.0.30320.0
ASP.NET and Web Tools 2017

ASP.NET Web Frameworks and Tools 2017   5.2.50303.0
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   15.0.30209.0
Azure App Service Tools v3.0.0

Common Azure Tools   1.9
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

GitHub.VisualStudio   2.2.0.8
A Visual Studio Extension that brings the GitHub Flow into Visual Studio.

JavaScript Language Service   2.0
JavaScript Language Service

KofePackagePackage Extension   1.0
KofePackagePackage Visual Studio Extension Detailed Info

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

NuGet Package Manager   4.1.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit http://docs.nuget.org/.

SQL Server Data Tools   15.1.61702.140
Microsoft SQL Server Data Tools

TypeScript   2.1.5.0
TypeScript tools for Visual Studio
```
**Steps to Reproduce**:
Run attached project.

**Expected Behavior**:
No null reference
**Actual Behavior**:
null reference.

Issue is cause by to ordering of the fields and when they get bound. We really have a 2nd pass after the readonly fields been initialized, to bind references that where initially uninitialised (hence Nothing) 

--------
@CyrusNajmabadi Quick overview of issue.

If I define the readonly fields as following it works as intended.
```vbnet
  ReadOnly BraceOpening As ParseDelegate = Rule(Name:=TokenKind.BraceOpening, Definition:=[Char]("{"c))
  ReadOnly BraceClosing As ParseDelegate = Rule(Name:=TokenKind.BraceClosing, Definition:=[Char]("}"c))
  ReadOnly Brace As ParseDelegate = Rule(Name:=TokenKind.Brace, Definition:=Choice(BraceOpening, BraceClosing))
```
But if I move the position of the `Brace` definition to above to other two, it doesn't work as intended.
```vbnet
  ReadOnly Brace As ParseDelegate = Rule(Name:=TokenKind.Brace, Definition:=Choice(BraceOpening, BraceClosing))
  ReadOnly BraceOpening As ParseDelegate = Rule(Name:=TokenKind.BraceOpening, Definition:=[Char]("{"c))
  ReadOnly BraceClosing As ParseDelegate = Rule(Name:=TokenKind.BraceClosing, Definition:=[Char]("}"c))

```
Because when `Brace` gets invoked, it is as `Definition:= Choice( Nothing, Nothing)`

**Comment by CyrusNajmabadi** [comment]
@AdamSpeight2008 It's not at all clear what the issue is that you're reporting.  Can you summarize in the OP?

**Comment by AdamSpeight2008** [comment]
@CyrusNajmabadi  Updated op.

**Comment by CyrusNajmabadi** [comment]
> but we should have an option to make this work.

Isn't the option "put your fields in the order you want them initialized in"?

**Comment by AdamSpeight2008** [comment]
Why? That would make that file ordering dependent. Just like overload and inherited methods.

When writing a read only field or property with an initializer, that reference other readonly fieald or property with and initialiser. It is (that reference) is being treated as if the field / property hasn't been initialised. 

Shouldn't the there be an option / attribute mark them as delay binding, till the other fields have initialised.

**Comment by CyrusNajmabadi** [comment]
> Why? That would make that file ordering dependent.

You want dependencies between your fields (since they reference each other).  If so, specifying your ordering is appropriate.

> Shouldn't the there be an option / attribute mark them as delay binding, till the other fields have initialised.

Sure.  Initialize them in a constructor :)   Or use Lazy.  Or use properties.

**Comment by AdamSpeight2008** [comment]
You'd still have the same issue, if I initialize them in a constructor. :smirk:

**Comment by CyrusNajmabadi** [comment]
Why?  Just initialize them in dependency order, and there's no issue.  If you want lazy binding, then use ```Lazy<T>```

**Comment by AdamSpeight2008** [comment]
@CyrusNajmabadi 
Is there a Diagnostic that advises informs the user that references will be null?

**Comment by CyrusNajmabadi** [comment]
Not that i know of.  but you're welcome to write an analyzer that looks for these sorts of things and lets someone know!

**Comment by AdamSpeight2008** [comment]
If if change the signature to `As Func(Of ParseDelegate)` it works at intended and expected.
```vbnet
ReadOnly Comma As Func(Of ParseDelegate)=Function() ...
```

**Comment by CyrusNajmabadi** [comment]
Yup.  Func's are another way to do delayed eval.

**Comment by gafter** [comment]
The compiler is specified to perform initializations in order, so this is not a bug. You can open a feature request at `https://github.com/dotnet/vblang` if you have something specific to suggest.
